test/client/submit.js

```diff
@@ -1210,5 +1212,189 @@ module.exports = function() {
+          next();
+        });
+
+        this.backend.use('apply', function(request, next) {
+          Object.assign(request.snapshot.m, request.op.m);
+          expect(request.op.m).to.have.property('username');
+          next();
+        });
+
+        this.backend.use('commit', function(request, next) {
+          expect(request.op.m).to.have.property('username');
+          next();
+        });
+
+        this.backend.use('afterWrite', function(request, next) {
+          expect(request.op.m).to.have.property('username');
+          next();
+        });
+
+        var subscriberCount = 10;
+        var subscriberOpCount = 10;
+
+        var metadatas = [];
+        for (var i = 0; i < subscriberCount; i++) {
+          metadatas[i] = {username: 'user-'+i};
+        }
+
+        var ops = [];
+        for (var i = 0; i < subscriberCount; i++) {
+          ops[i] = [];
+          for (var j = 0; j < subscriberOpCount; j++) {
+            ops[i].push({p: ['tricks '+i+' '+j], oi: 1});
+          }
+        }
+
+        var docs = [];
+
+        function submitOps() {
+          for (var j = 0; j < subscriberOpCount; j++) {
+            for (var i = 0; i < subscriberCount; i++) {
+              var doc = docs[i];
+              doc.submitOp([ops[i][j]], {source: 'src-'+i}, errorHandler(doneAfter));
+            }
+          }
+        }
+
+        function validateAndDone() {
+          var firstDoc = docs[0];
+          // validate that all documents across connections are in sync
+          for (var i = 1; i < subscriberCount; i++) {
+            var doc = docs[i];
+            expect(doc.data).eql(firstDoc.data);
+          }
+          done();
+        };
+
+        var submitOpsAfter = util.callAfter(subscriberCount - 1, submitOps);
+        var doneAfter = util.callAfter((subscriberCount * subscriberCount * subscriberOpCount) - 1, validateAndDone);
+
+        function getDoc(callback) {
+          var thisDoc = this;
+          thisDoc.fetch(function() {
+            if (!thisDoc.data) {
+              return thisDoc.create({}, function() {
+                thisDoc.subscribe(callback);
+              });
+            }
+            thisDoc.subscribe(callback);
+          });
+        }
+
+        for (var i = 0; i < subscriberCount; i++) {
+          var metadata = metadatas[i];
+
+          var connection = this.backend.connect(undefined, Object.assign({}, metadata));
+          connection.__test_metadata = Object.assign({}, metadata);
+          connection.__test_id = i;
+
+          connection.on('connected', function() {
+            var thisConnection = this;
+
+            expect(thisConnection.agent.custom).eql(thisConnection.__test_metadata);
+
+            thisConnection.doc = docs[thisConnection.__test_id] = thisConnection.get('dogs', 'fido');
+
+            thisConnection.doc.on('op', function(op, source, src, context) {
+              if (!src || !context) { // If I am the source there is no metadata to check
```

Comment:
Why don't we send metadata back to the original client? The server might add context the sender wants (and which remote clients will get).

What's your use-case? Does it work without the original op submitter getting their metadata?

We could add the meta to the op acknowledgement.